Add minimal cpu affinity

carlossvg opened the pull request.

carlos-apex committed:
Add minimal_cpu_affinity package

shuhaowu commented on minimal_cpu_affinity/minimal_cpu_affinity_main_thread.cpp:
```diff
@@ -0,0 +1,81 @@
+
+  auto node = std::make_shared<MinimalPublisher>();
+
+  int cpu_id = 1;
```
In the _mask.cpp file you had a comment that says "Pin the main thread to CPU 2 and 3". Can you put a similar comment here as well to guide the reader?

carlossvg replied:
done

shuhaowu reviewed (approved):
Should probably also put a note about how this example is only demoing the affinity functions and is not fully "RT".

Also might want to make a note on why you wand to do this in the README?

carlossvg commented:
I added some documentation. I was thinking about adding an additional note or sections explaining that it is possible to use [taskset](https://www.man7.org/linux/man-pages/man1/taskset.1.html) to set the CPU affinity for the whole process. What do you think? Same question for the minimal_scheduling package and [chrt](https://www.man7.org/linux/man-pages/man1/chrt.1.html).

shuhaowu commented:
Maybe just mention it and link to the docs? I tend to see the command line tools to be something you use for adhoc testing. For a "rt" application I would imagine it would do this by itself via the code you have here.

awischnewski commented:
I'll just add the thoughts from here #7 which still apply here: 
- [x] The ps examples should use `ps -C minimal_cpu_aff -L -o tid,comm,rtprio,cls,psr`, seems to be a copy paste error.
- [x] It might be helpful to add a hint that htop uses a different CPU numbering (starting from 1 instead of 0), even though that might be less important now as the docs already explain how to do it with ps. 
- [ ] What do you think about adding a hint to how to use the `isolcpus` option in the bootloader? This is a nice feature in conjunction with cpu affinity and helps to get descent RT performance with little effort. (https://www.skenz.it/cs/kernel_boot_parameters for ubuntu)

carlossvg commented:
>The ps examples should use ps -C minimal_cpu_aff -L -o tid,comm,rtprio,cls,psr, seems to be a copy paste error.

Thanks, I will update this 

>It might be helpful to add a hint that htop uses a different CPU numbering (starting from 1 instead of 0), even though that might be less important now as the docs already explain how to do it with ps.

Makes sense

>What do you think about adding a hint to how to use the isolcpus option in the bootloader? This is a nice feature in conjunction with cpu affinity and helps to get descent RT performance with little effort. (https://www.skenz.it/cs/kernel_boot_parameters for ubuntu)

I was thinking about extending the example to show CPU isolation. Documenting this is already in our ToDo list so maybe it's worth to add that document first and simply refer to it in the example. Apart from isolcpus I would example how to do it with cgorups and CPUSET.

carlossvg commented:
@shuhaowu @awischnewski Sorry for the delay. I've been busy in the last weeks. I addressed your comments and I added some recordings to the documentation https://ros-realtime.github.io/ros2-realtime-examples/minimal_cpu_affinity/

I think it would interesting to show some examples using CPU isolation but I would do it in a follow-up issue.
